Student Performance Analyzer — NumPy Project

In [1]:
#.Import the libraries
import pandas as pd
import numpy as np

In [2]:
#Task 1 — Load and Convert the Dataset
df=pd.read_csv(r"student_performance.csv")
df.head()

,Student_ID,Student,Math,Science,English,Computer,Attendance
0,1,Aarav,85,78,92,88,95
1,2,Diya,72,81,75,79,88
2,3,Arjun,91,89,84,95,97
3,4,Ananya,65,70,68,72,76
4,5,Rohan,88,94,90,91,93


In [3]:
#convert numerical data into numpy array
data = df[['Student_ID', 'Math', 'Science', 'English', 'Computer', 'Attendance']].to_numpy()
print(data)

[[ 1 85 78 92 88 95]
 [ 2 72 81 75 79 88]
 [ 3 91 89 84 95 97]
 [ 4 65 70 68 72 76]
 [ 5 88 94 90 91 93]
 [ 6 76 85 79 82 89]
 [ 7 95 92 88 96 98]
 [ 8 69 74 81 70 82]
 [ 9 82 77 85 89 91]
 [10 90 86 93 94 96]
 [11 58 65 62 68 71]
 [12 79 83 87 80 86]
 [13 87 91 78 90 94]
 [14 73 69 76 75 79]
 [15 94 88 91 97 99]
 [16 67 72 70 65 74]
 [17 81 80 84 86 90]
 [18 89 93 88 92 95]
 [19 61 68 64 70 69]
 [20 84 87 90 85 92]]


In [4]:
#Task 2 — Explore the Dataset
print("Shape:", data.shape)
print("Dimensions:", data.ndim)
print("Number of Students:", data.shape[0])
print("Total No. of Elements:", data.size)

Shape: (20, 6)
Dimensions: 2
Number of Students: 20
Total No. of Elements: 120


In [5]:
# Task 3 — Indexing and Slicing

print("First Student:", data[0])              # First student
print("Last Student:", data[-1])              # Last student

First Student: [ 1 85 78 92 88 95]
Last Student: [20 84 87 90 85 92]


In [6]:
print("First 5 Students:",data[:5])           # First 5 students

First 5 Students: [[ 1 85 78 92 88 95]
 [ 2 72 81 75 79 88]
 [ 3 91 89 84 95 97]
 [ 4 65 70 68 72 76]
 [ 5 88 94 90 91 93]]


In [7]:
print("Last 5 Students:",data[-5:])           # Last 5 students

Last 5 Students: [[16 67 72 70 65 74]
 [17 81 80 84 86 90]
 [18 89 93 88 92 95]
 [19 61 68 64 70 69]
 [20 84 87 90 85 92]]


In [8]:
print("Math Scores:", data[:, 1])             # Math scores
print("Attendance:", data[:, 5])              #Attendance

Math Scores: [85 72 91 65 88 76 95 69 82 90 58 79 87 73 94 67 81 89 61 84]
Attendance: [95 88 97 76 93 89 98 82 91 96 71 86 94 79 99 74 90 95 69 92]


In [9]:
print("5th to 10th Students:",data[4:10])     # 5th to 10th students

5th to 10th Students: [[ 5 88 94 90 91 93]
 [ 6 76 85 79 82 89]
 [ 7 95 92 88 96 98]
 [ 8 69 74 81 70 82]
 [ 9 82 77 85 89 91]
 [10 90 86 93 94 96]]


In [10]:
#Task 4 — Basic Statistical Analysis

scores = data[:, 1:]

print("Mean:", np.mean(scores, axis=0))
print("Median:", np.median(scores, axis=0))
print("Minimum:", np.min(scores, axis=0))
print("Maximum:", np.max(scores, axis=0))
print("Standard Deviation:", np.std(scores, axis=0))
print("Variance:", np.var(scores, axis=0))

Mean: [79.3  81.1  81.25 83.2  87.7 ]
Median: [81.5 82.  84.  85.5 90.5]
Minimum: [58 65 62 65 69]
Maximum: [95 94 93 97 99]
Standard Deviation: [10.86784247  8.87073841  9.2405357   9.98799279  9.12195155]
Variance: [118.11    78.69    85.3875  99.76    83.21  ]


In [11]:
# Task 5 — Student Performance Classification

# Calculate average score of each student
subject_scores = data[:, 1:5]
average_score = np.mean(subject_scores, axis=1)

# Classify students
performance = np.where(average_score >= 80, "Excellent",
              np.where(average_score >= 60, "Good",
              np.where(average_score >= 40, "Average","Needs Improvement")))
print("Average Scores:", average_score)
print("Performance:", performance)

Average Scores: [85.75 76.75 89.75 68.75 90.75 80.5  92.75 73.5  83.25 90.75 63.25 82.25
 86.5  73.25 92.5  68.5  82.75 90.5  65.75 86.5 ]
Performance: ['Excellent' 'Good' 'Excellent' 'Good' 'Excellent' 'Excellent' 'Excellent'
 'Good' 'Excellent' 'Excellent' 'Good' 'Excellent' 'Excellent' 'Good'
 'Excellent' 'Good' 'Excellent' 'Excellent' 'Good' 'Excellent']


In [12]:
# Task 6 — Attendance and Performance Analysis

# Average attendance
average_attendance = np.mean(data[:, 5])
print("Average Attendance:", average_attendance)

# Students with above-average attendance
above_average = data[:, 5] > average_attendance
print("Above Average Attendance:", above_average)

# Performance of students with above-average attendance
high_attendance_score = average_score[above_average]
print("Average Performance of High Attendance Students:",np.mean(high_attendance_score))

# Students with below-average attendance
below_average = data[:, 5] <= average_attendance

# Performance of students with below-average attendance
low_attendance_score = average_score[below_average]
print("Average Performance of Low Attendance Students:",np.mean(low_attendance_score))

Average Attendance: 87.7
Above Average Attendance: [ True  True  True False  True  True  True False  True  True False False
  True False  True False  True  True False  True]
Average Performance of High Attendance Students: 86.84615384615384
Average Performance of Low Attendance Students: 70.75


In [13]:
# Task 7 — Top and Bottom Performers

# Highest-performing student
top_index = np.argmax(average_score)
print("Top Student ID:", data[top_index, 0])
print("Top Student Score:", average_score[top_index])

Top Student ID: 7
Top Student Score: 92.75


In [14]:
# Lowest-performing student
bottom_index = np.argmin(average_score)
print("Lowest Student ID:", data[bottom_index, 0])
print("Lowest Student Score:", average_score[bottom_index])

Lowest Student ID: 11
Lowest Student Score: 63.25


In [15]:
# Top 3 performers
top_3_indices = np.argsort(average_score)[-3:][::-1]
print("Top 3 Students:")
print("Student IDs:", data[top_3_indices, 0])
print("Scores:", average_score[top_3_indices])

Top 3 Students:
Student IDs: [ 7 15 10]
Scores: [92.75 92.5  90.75]


In [16]:
# Bottom 3 performers
bottom_3_indices = np.argsort(average_score)[:3]
print("Bottom 3 Students:")
print("Student IDs:", data[bottom_3_indices, 0])
print("Scores:", average_score[bottom_3_indices])

Bottom 3 Students:
Student IDs: [11 19 16]
Scores: [63.25 65.75 68.5 ]


In [17]:
# Task 8 — Subject-wise Performance Analysis

scores = data[:, 1:5]
subjects = np.array(["Math", "Science", "English", "Computer"])

# Average of each subject
subject_means = np.mean(scores, axis=0)
print("Subject Means:", subject_means)

Subject Means: [79.3  81.1  81.25 83.2 ]


In [18]:
# Highest average subject
highest_subject_index = np.argmax(subject_means)
print("Highest Average Subject:", subjects[highest_subject_index])
print("Average Score:", subject_means[highest_subject_index])

Highest Average Subject: Computer
Average Score: 83.2


In [19]:
# Lowest average subject
lowest_subject_index = np.argmin(subject_means)
print("Lowest Average Subject:", subjects[lowest_subject_index])
print("Average Score:", subject_means[lowest_subject_index])

# Overall average
overall_average = np.mean(scores)
print("Overall Average:", overall_average)

Lowest Average Subject: Math
Average Score: 79.3
Overall Average: 81.2125


In [20]:
# Students scoring 80 or above
count_above_80 = np.sum(scores >= 80, axis=0)
print("Students scoring 80 or above:", count_above_80)

Students scoring 80 or above: [11 12 12 13]


In [21]:
# Task 9 — Overall Performance Score

scores = data[:, 1:5]

# Calculate overall score for each student
overall_score = np.sum(scores, axis=1) / 4
print("Overall Score:", overall_score)

Overall Score: [85.75 76.75 89.75 68.75 90.75 80.5  92.75 73.5  83.25 90.75 63.25 82.25
 86.5  73.25 92.5  68.5  82.75 90.5  65.75 86.5 ]


In [23]:
# Class average
class_average = np.mean(overall_score)
print("Class Average:", class_average)

# Top-performing student
top_index = np.argmax(overall_score)
print("Top Student ID:", data[top_index, 0])
print("Top Overall Score:", overall_score[top_index])

Class Average: 81.2125
Top Student ID: 7
Top Overall Score: 92.75


In [24]:
# Number of students above class average
above_class_average = overall_score > class_average
print("Students above class average:", np.sum(above_class_average))

Students above class average: 12


In [25]:
# Task 10 — Final Insights

print("Total Students:", data.shape[0])

print("Class Average:", np.mean(overall_score))

print("Highest Score:", np.max(overall_score))

print("Lowest Score:", np.min(overall_score))

excellent_students = np.sum(performance == "Excellent")
good_students = np.sum(performance == "Good")

print("Excellent Students:", excellent_students)
print("Good Students:", good_students)

highest_subject = subjects[np.argmax(subject_means)]
lowest_subject = subjects[np.argmin(subject_means)]

print("Highest Average Subject:", highest_subject)
print("Lowest Average Subject:", lowest_subject)

above_average_count = np.sum(overall_score > np.mean(overall_score))

print("Students Above Class Average:", above_average_count)

Total Students: 20
Class Average: 81.2125
Highest Score: 92.75
Lowest Score: 63.25
Excellent Students: 13
Good Students: 7
Highest Average Subject: Computer
Lowest Average Subject: Math
Students Above Class Average: 12


The analysis of 20 students shows an overall class average of 81.21. Most students performed well, with 13 students classified as Excellent and 7 as Good. Computer had the highest average performance among the subjects, while Math had the lowest. Additionally, 12 students scored above the overall class average, indicating generally strong academic performance in the dataset.